### PR TITLE
[wni] Enable SSH networking for windows node

### PR DESCRIPTION
**Pull request by PratikMahajan**
This PR opens up the SSH port for local IP of the user and installs OpenSSH server on the windows node for Azure VMs. This PR includes e2e tests for testing OpenSSH connections on Azure VMs

### CODE REVIEW

**Comment by PratikMahajan** [comment]
```
[mahajan@localhost windows-node-installer]$ go test -run=TestCreateVM ./test/e2e/... -v
=== RUN   TestCreateVM
2020/03/13 8:53:15 kubeconfig source: /home/mahajan/projects/aws-infra/extras/pmahajan-az07/auth/kubeconfig
=== RUN   TestCreateVM/check_if_required_security_rules_are_present
=== RUN   TestCreateVM/check_if_ansible_is_able_to_ping_on_the_WinRmHttps_port
=== RUN   TestCreateVM/check_if_container_logs_port_is_open_in_Windows_firewall
=== RUN   TestCreateVM/check_if_SSH_connection_is_available
--- PASS: TestCreateVM (12.65s)
    --- PASS: TestCreateVM/check_if_required_security_rules_are_present (0.85s)
    --- PASS: TestCreateVM/check_if_ansible_is_able_to_ping_on_the_WinRmHttps_port (3.44s)
    --- PASS: TestCreateVM/check_if_container_logs_port_is_open_in_Windows_firewall (7.13s)
    --- PASS: TestCreateVM/check_if_SSH_connection_is_available (0.45s)
PASS
ok      github.com/openshift/windows-machine-config-bootstrapper/tools/windows-node-installer/test/e2e  12.659s
[mahajan@localhost windows-node-installer]$ 
```

File: tools/windows-node-installer/pkg/cloudprovider/azure/vm.go [hunk 1]
```diff
@@ -50,6 +50,12 @@ const (
 	vnetRuleName = "vnet_traffic"
 	// winUser is the user used to login into the instance.
 	winUser = "core"
+	// sshPort is the SSH port
+	sshPort = "22"
```
Comment (ravisantoshgudimetla) [inline tools/windows-node-installer/pkg/cloudprovider/azure/vm.go]: If you make these public, won't you be able to re-use them in the test code? or are you not importing this package in the test suite?

File: tools/windows-node-installer/test/e2e/azure_test.go [hunk 2]
```diff
@@ -354,8 +363,29 @@ func testAnsiblePing(t *testing.T) {
...
 func testAzureInstancesFirewallRule(t *testing.T) {
+	var err error
+	windowsVM, err = getAzureCreds()
+	require.NoErrorf(t, err, "failed to obtain credentials for %s", credentials.GetInstanceId())
+	testInstanceFirewallRule(t)
+}
+
+func getAzureCreds() (windowsVM types.WindowsVM, err error) {
+	w := &types.Windows{}
+	ipAddress, password, err := getIpPass(credentials.GetInstanceId())
```
Comment (ravisantoshgudimetla) [inline tools/windows-node-installer/test/e2e/azure_test.go]: I think you should early if you got an error here

File: tools/windows-node-installer/test/e2e/azure_test.go [hunk 3]
```diff
@@ -354,8 +363,29 @@ func testAnsiblePing(t *testing.T) {
 
 // testAzureFirewallRule asserts if the created instance has firewall rule that opens container logs port.
 func testAzureInstancesFirewallRule(t *testing.T) {
+	var err error
+	windowsVM, err = getAzureCreds()
+	require.NoErrorf(t, err, "failed to obtain credentials for %s", credentials.GetInstanceId())
+	testInstanceFirewallRule(t)
+}
+
+func getAzureCreds() (windowsVM types.WindowsVM, err error) {
```
Comment (ravisantoshgudimetla) [inline tools/windows-node-installer/test/e2e/azure_test.go]: please add comments for the function. Aren't we returning windowsVM here, why is it called `getAzureCreds`?

File: tools/windows-node-installer/test/e2e/azure_test.go [hunk 4]
```diff
@@ -354,8 +363,29 @@ func testAnsiblePing(t *testing.T) {
...
 	ipAddress, password, err := getIpPass(credentials.GetInstanceId())
 	require.NoErrorf(t, err, "failed to obtain credentials for %s", credentials.GetInstanceId())
 	credentials = types.NewCredentials(credentials.GetInstanceId(), ipAddress, password, azureUser)
-	testInstanceFirewallRule(t)
+	config := &ssh.ClientConfig{
+		User:            credentials.GetUserName(),
+		Auth:            []ssh.AuthMethod{ssh.Password(credentials.GetPassword())},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+	_, err = ssh.Dial("tcp", credentials.GetIPAddress()+":22", config)
```
Comment (ravisantoshgudimetla) [inline tools/windows-node-installer/test/e2e/azure_test.go]: Can you check if the object if not nil or empty? Perhaps there is nothing wrong in doing a windowsVM.Run(some comman) to prove it out?

File: tools/windows-node-installer/test/e2e/azure_test.go [hunk 5]
```diff
@@ -354,8 +363,29 @@ func testAnsiblePing(t *testing.T) {
...
+
+func getAzureCreds() (windowsVM types.WindowsVM, err error) {
+	w := &types.Windows{}
+	ipAddress, password, err := getIpPass(credentials.GetInstanceId())
+	credentials = types.NewCredentials(credentials.GetInstanceId(), ipAddress, password, azureUser)
+	w.Credentials = credentials
+	return w, err
+}
+
+func testAzureSSHConnection(t *testing.T) {
```
Comment (ravisantoshgudimetla) [inline tools/windows-node-installer/test/e2e/azure_test.go]: Please add comments.

File: tools/windows-node-installer/test/e2e/azure_test.go (hunk 3 above)
Comment (akhil-rane) [reply tools/windows-node-installer/test/e2e/azure_test.go]: Rename 'getAzureCreds' to something like 'getAzureWindowsVM'

File: tools/windows-node-installer/pkg/cloudprovider/azure/vm.go (hunk 1 above)
Comment (PratikMahajan) [reply tools/windows-node-installer/pkg/cloudprovider/azure/vm.go]: we haven't imported the package in test suite. Was just following the convention that we followed in the code

File: tools/windows-node-installer/test/e2e/azure_test.go (hunk 2 above)
Comment (PratikMahajan) [reply tools/windows-node-installer/test/e2e/azure_test.go]: will do it

File: tools/windows-node-installer/pkg/cloudprovider/azure/vm.go (hunk 1 above)
Comment (aravindhp) [reply tools/windows-node-installer/pkg/cloudprovider/azure/vm.go]: This is also declared in the [AWS](https://github.com/openshift/windows-machine-config-bootstrapper/blob/master/tools/windows-node-installer/pkg/cloudprovider/aws/ec2.go#L43) pkg. Please have a pre-commit that moves it to a common pkg so that it can be used here.

File: tools/windows-node-installer/test/e2e/azure_test.go [hunk 6]
```diff
@@ -47,6 +48,12 @@ const (
 	ruleAction = "Allow"
 	// azureUser used to access the windows instance
 	azureUser = "core"
+	// sshPort is the SSH port
+	sshPort = "22"
```
Comment (aravindhp) [inline tools/windows-node-installer/test/e2e/azure_test.go]: Same comment here about sharing it between the test packages of AWS and Azure. You can have one pre-commit where you make it common for both test and regular packages.

**Comment by aravindhp** [comment]
I also not sure what you mean by `wni installer IP` in your PR message. This is just the local IP address of the user.

File: tools/windows-node-installer/README.md [hunk 7]
```diff
@@ -107,7 +107,7 @@ are run. In addition windows-node-installer.json and winworker-* files need to b
 
 To run the e2e change the directory to the `tools\windows-node-installer` of this repo and run the following commands:
 ```bash
-export ARTIFACT_DIRECTORY=$(pwd)
+export ARTIFACT_DIR=$(pwd)
```
Comment (ravisantoshgudimetla) [inline tools/windows-node-installer/README.md]: Why is this needed?

File: tools/windows-node-installer/pkg/cloudprovider/aws/ec2.go [hunk 8]
```diff
@@ -40,7 +40,7 @@ const (
 	awsPasswordDataTimeOut = 15 * time.Minute
 	// sshPort to access the OpenSSH server installed on the windows node. This is needed
 	// for our CI testing.
-	sshPort = 22
+	sshPort = types.SshPort
```
Comment (ravisantoshgudimetla) [inline tools/windows-node-installer/pkg/cloudprovider/aws/ec2.go]: This is unnecessary change again. Either update the entire code to use `types.SshPort` don't bring this change in.

File: tools/windows-node-installer/pkg/cloudprovider/azure/vm.go [hunk 9]
```diff
@@ -50,6 +50,12 @@ const (
 	vnetRuleName = "vnet_traffic"
 	// winUser is the user used to login into the instance.
 	winUser = "core"
+	// sshPort is the SSH port
+	sshPort = string(types.SshPort)
```
Comment (ravisantoshgudimetla) [inline tools/windows-node-installer/pkg/cloudprovider/azure/vm.go]: Why this explicit type casting? Ideally you should have type casting within the function by creating another var

File: tools/windows-node-installer/test/e2e/common.go [hunk 10]
```diff
@@ -16,6 +16,8 @@ import (
 const (
 	// winRMPort is the secure WinRM port
 	winRMPort = "5986"
+	// sshPort is the SSH port
+	sshPort = "22"
```
Comment (ravisantoshgudimetla) [inline tools/windows-node-installer/test/e2e/common.go]: nit: If you made `sshPort` public in the code you could have directly imported?

File: tools/windows-node-installer/pkg/cloudprovider/aws/ec2.go (hunk 8 above)
Comment (aravindhp) [reply tools/windows-node-installer/pkg/cloudprovider/aws/ec2.go]: I agree with @ravisantoshgudimetla. I had clearly called out that these sort of changes need to be done in a separate pre-commit to the main one.

File: tools/windows-node-installer/pkg/types/types.go [hunk 11]
```diff
@@ -28,6 +28,8 @@ const (
 	remotePowerShellCmdPrefix = "powershell.exe -NonInteractive -ExecutionPolicy Bypass "
 	// winRMPort is port used for WinRM communication
 	winRMPort = 5986
+	// sshPort is the SSH port
```
Comment (aravindhp) [inline tools/windows-node-installer/pkg/types/types.go]: Please fix the comment.

File: tools/windows-node-installer/test/e2e/common.go (hunk 10 above)
Comment (aravindhp) [reply tools/windows-node-installer/test/e2e/common.go]: I think the constants are not being shared across the main packages and the test code.

File: tools/windows-node-installer/test/e2e/azure_test.go [hunk 12]
```diff
@@ -106,6 +105,7 @@ func TestCreateVM(t *testing.T) {
 	t.Run("check if required security rules are present", testRequiredRules)
 	t.Run("check if ansible is able to ping on the WinRmHttps port", testAnsiblePing)
 	t.Run("check if container logs port is open in Windows firewall", testAzureInstancesFirewallRule)
+	t.Run("check if SSH connection is available", testAzureSSHConnection)
```
Comment (aravindhp) [inline tools/windows-node-installer/test/e2e/azure_test.go]: The e2e test should be part of the commit where you are introducing the feature.

File: tools/windows-node-installer/README.md (hunk 7 above)
Comment (PratikMahajan) [reply tools/windows-node-installer/README.md]: in the code its referenced to as ARTIFACT_DIR whereas in readme its given as ARTIFACT_DIRECTORY

Comment (aravindhp) [reply tools/windows-node-installer/README.md]: Sure, but you should not fix that as part of this PR.

**Comment by ravisantoshgudimetla** [comment]
Can the hold be cancelled @aravindhp?

**Comment by aravindhp** [comment]
@ravisantoshgudimetla I placed a hold because @PratikMahajan seems to have not followed our new guidelines. It is up to him to remove it.
